### PR TITLE
Run style checks in Azure Pipelines

### PR DESCRIPTION
**Pull request by santisoler**
Run style checks using flake8 and some of its plugins. Add a new
`.flake8` configuration file to tweak the style checks. Add a new
`check` target in the `Makefile`. Create a new pipeline in Azure to
automatically run style checks on every PullRequest. Add new required
packages to `environment_test.yml` and to a new
`requirements_style.txt`.

### CODE REVIEW

**Comment by santisoler** [comment]
I expect the style checks to fail big time now, but sets the basis for solving those issues and also choose which style checks we would like to enforce in SimPEG.

One major thing I haven't added is [`pep8-naming`](https://pypi.org/project/pep8-naming/), a package and `flake8` plugin that checks that variable, functions and classes names follow PEP8. I wouldn't add it right away in SimPEG since it will complain a lot now, but I'm mentioning it to keep record that in the long run would be nice to have it.

**Comment by jcapriot** [comment]
I'm still surprised there's a Makefile left over in this repo, it is barely used for anything, and there's no compiled codes in here anyways that would make it that necessary.

**Comment by jcapriot** [comment]
@santisoler looks like there's an error in the azure-pipelines file with a bad name for the style check stage https://dev.azure.com/simpeg/simpeg/_build/results?buildId=2110&view=results. Probably needs to be renamed "Style_Check". You can always explicitly set it's displayname though.

**Comment by santisoler** [comment]
> I'm still surprised there's a Makefile left over in this repo, It is barely used for anything, and there's no compiled codes in here anyways that would make it that necessary.

I saw @rowanc1's hand in that Makefile. I imagined it's not being heavily used in SimPEG. I usually love to see a Makefile in the repo (even if the repo doesn't involve compilation steps) because it provides an easy way to standardize processes, document them and make them quickly available. For example, instead of adding instructions in the README or in the Contributors Guide on how to run every style check, one could just say "run `make check`". We use this approach in every Fatiando package (see https://github.com/fatiando/verde/blob/main/Makefile) and we use them in our [Contributing Guidelines](https://github.com/fatiando/community/blob/main/CONTRIBUTING.md#code-style). One can also use it in CI to run tests and code styles (like I'm doing here). So if you want to change something about how the code style gets done, you just need to edit the `Makefile` and leave the azure pipeline file intact.

**Comment by jcapriot** [comment]
We probably want to (for now) allow these tests to fail and still continue with unit tests.

File: environment_test.yml [hunk 1]
```diff
@@ -36,3 +36,10 @@ dependencies:
   - pyvista
   - pip
   - python-kaleido
+  # Linters and code style
+  - black
```
Comment (jcapriot) [inline environment_test.yml:40]: make sure that this version matches the current version in the .pre-commit.yml file. (I think 21.7b0)

File: environment_test.yml [hunk 2]
```diff
@@ -36,3 +36,10 @@ dependencies:
...
+  - flake8
+  - flake8-bugbear
+  - flake8-builtins
+  - flake8-mutable
+  - flake8-rst-docstrings
```
Comment (jcapriot) [inline environment_test.yml:45]: we might also want to use `flake8-docstrings` for numpy style documentation checking
https://pypi.org/project/flake8-docstrings/

**Comment by santisoler** [comment]
> We probably want to (for now) allow these tests to fail and still continue with unit tests.

Sure. Let's just merge this PR and fix all the style issues in separate PRs.

**Comment by rowanc1** [comment]
Looks like the `make graphs` is still in there as well, which made some figures for various papers about the inheritance structure. Exciting to see these new improvements!!

**Comment by santisoler** [comment]
Thanks @jcapriot for all the changes. I just grouped the target files into a single variable in the Makefile to avoid repetition and to make it easier to change in the future. Feel free to revert it if you don't like it. From my point this PR is ready, so merge it whenever you feel the same.

**Comment by jcapriot** [comment]
Awesome, I think the only thing that looks obvious to turn off is the documentation style checking in tests

**Comment by santisoler** [comment]
> Awesome, I think the only thing that looks obvious to turn off is the documentation style checking in tests

Makes sense. We'll never render docstrings of test functions in Sphinx, so we don't care about their format, they just need to be informative when reading them in plain text.

**Pull request merged**